framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -2448,8 +2465,120 @@ protected boolean isTypeCastSafe(AnnotatedTypeMirror castType, AnnotatedTypeMirr
+                Set<AnnotationMirror> exprLower =
+                        AnnotatedTypes.findEffectiveLowerBoundAnnotations(
+                                qualifierHierarchy, exprType);
+                Set<AnnotationMirror> castUpper =
+                        AnnotatedTypes.findEffectiveAnnotations(qualifierHierarchy, castType);
+                Set<AnnotationMirror> exprUpper =
+                        AnnotatedTypes.findEffectiveAnnotations(qualifierHierarchy, exprType);
+
+                // return SAFE only it satisfies the below condition
+                if (qualifierHierarchy.isSubtype(exprLower, castLower)
+                        && qualifierHierarchy.isSubtype(exprUpper, castUpper)) {
+                    return CastSafeKind.SAFE;
+                }
+                return CastSafeKind.WARNING;
+            }
+        }
+
         AnnotatedTypeMirror exprTypeWidened = atypeFactory.getWidenedType(exprType, castType);
-        return qualifierHierarchy.isSubtype(exprTypeWidened.getEffectiveAnnotations(), castAnnos);
+        boolean result =
+                qualifierHierarchy.isSubtype(exprTypeWidened.getEffectiveAnnotations(), castAnnos);
+
+        if (result) {
+            return CastSafeKind.SAFE;
+        } else if (checkCastElementType) {
+            // when the flag is enabled, and it is not an upcast, return a warning
+            return CastSafeKind.WARNING;
+        } else {
+            return CastSafeKind.NOT_UPCAST;
+        }
+    }
+
+    /**
+     * Determine the typecast is downcast or not. If it is, further determine it can be statically
+     * verified or not.
+     *
+     * @param castType annotated type of the cast
+     * @param exprType annotated type of the casted expression
+     * @return Can return SAFE, NOT_DOWNCAST or WARNING.
+     */
+    protected CastSafeKind isSafeDowncast(
+            AnnotatedTypeMirror castType, AnnotatedTypeMirror exprType) {
+        Set<AnnotationMirror> castAnnos = castType.getEffectiveAnnotations();
+        Set<AnnotationMirror> exprAnnos = exprType.getEffectiveAnnotations();
+        QualifierHierarchy qualifierHierarchy = atypeFactory.getQualifierHierarchy();
+
+        if (!qualifierHierarchy.isComparable(castAnnos, exprAnnos)) { // exists an incomparable cast
+            return CastSafeKind.NOT_DOWNCAST;
+        }
+
+        // check if the downcast can be statically verified
+        final TypeKind castTypeKind = castType.getKind();
+
+        if (castTypeKind == TypeKind.DECLARED) {
+            // Don't issue an error if the annotations are equivalent to the qualifier upper bound
+            // of the type.
+            AnnotatedDeclaredType castDeclared = (AnnotatedDeclaredType) castType;
+            Set<AnnotationMirror> bounds =
+                    atypeFactory.getTypeDeclarationBounds(castDeclared.getUnderlyingType());
+
+            if (AnnotationUtils.areSame(castDeclared.getAnnotations(), bounds)) {
```

Comment:
Must declared type and bounds be the same? How about declared type is the subtype of bounds and also satisfies that this subtype is also a subtype of expr type?

Reply:
Suppose the (upper) bound of the type `classA` is `@Super`, the castDeclared's annotation (use annotation of the class) is `@SubA` (subtype of `@Super`). When successfully downcasting from object to this java type `classA`, we get additional information that this object must have the annotation which is upper bounded by `@Super`, but we can't get more precise information about the annotation type, it could be `@SubA` or `@SubB` or `SubC`.